ICE: `Invalid float op Neg`

<!--
[31mICE[0m: Rustc ./ice.rs '' 'error: internal compiler error: /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/compiler/rustc_const_eval/src/interpret/operator.rs:498:26: Invalid float op Neg', 'error: internal compiler error: /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/compiler/rustc_const_eval/src/interpret/operator.rs:498:26: Invalid float op Neg'
File: /tmp/im/ice.rs
-->

snippet:
````rust
fn main() {
    let _ = -(-0.0f16);
}


````

Version information
````
rustc 1.80.0-nightly (378a43a06 2024-05-01)
binary: rustc
commit-hash: 378a43a06510f3e3a49c69c8de71745e6a884048
commit-date: 2024-05-01
host: x86_64-unknown-linux-gnu
release: 1.80.0-nightly
LLVM version: 18.1.4

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: the type `f16` is unstable
 --> /tmp/icemaker_global_tempdir.LUs0yTDS2O3R/rustc_testrunner_tmpdir_reporting.laee65vXPGyd/mvce.rs:2:16
  |
2 |     let _ = -(-0.0f16);
  |                ^^^^^^
  |
  = note: see issue #116909 <https://github.com/rust-lang/rust/issues/116909> for more information
  = help: add `#![feature(f16)]` to the crate attributes to enable
  = note: this compiler was built on 2024-05-01; consider upgrading it if it is out of date

error: internal compiler error: /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/compiler/rustc_const_eval/src/interpret/operator.rs:498:26: Invalid float op Neg
 --> /tmp/icemaker_global_tempdir.LUs0yTDS2O3R/rustc_testrunner_tmpdir_reporting.laee65vXPGyd/mvce.rs:1:1
  |
1 | fn main() {
  | ^^^^^^^^^

thread 'rustc' panicked at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/compiler/rustc_const_eval/src/interpret/operator.rs:498:26:
Box<dyn Any>
stack backtrace:
   0:     0x75fcab3ba035 - std::backtrace_rs::backtrace::libunwind::trace::hda47b82a166812ce
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
   1:     0x75fcab3ba035 - std::backtrace_rs::backtrace::trace_unsynchronized::hfdfc54f1e69a0c51
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x75fcab3ba035 - std::sys_common::backtrace::_print_fmt::h793b3f7571d13dfd
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x75fcab3ba035 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hbc9f1fbbb950dc64
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x75fcab40929b - core::fmt::rt::Argument::fmt::h616c7e5e48e69035
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/core/src/fmt/rt.rs:165:63
   5:     0x75fcab40929b - core::fmt::write::h273cb7cce9dfa202
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/core/src/fmt/mod.rs:1157:21
   6:     0x75fcab3aebdf - std::io::Write::write_fmt::h945d30311b10e894
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/io/mod.rs:1832:15
   7:     0x75fcab3b9e0e - std::sys_common::backtrace::_print::hcdebe37872d00154
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x75fcab3b9e0e - std::sys_common::backtrace::print::h51bd45a216cc9df6
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x75fcab3bc779 - std::panicking::default_hook::{{closure}}::hb6bc4eb0d96a4bfd
  10:     0x75fcab3bc4bd - std::panicking::default_hook::h557ff9f57d861b48
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/panicking.rs:298:9
  11:     0x75fca7c3c6bb - std[8fa947ca3cb89c52]::panicking::update_hook::<alloc[4c35cc17dd12f569]::boxed::Box<rustc_driver_impl[f52ddb602ce7d3c0]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x75fcab3bce76 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hf5d9c19b8a775637
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/alloc/src/boxed.rs:2036:9
  13:     0x75fcab3bce76 - std::panicking::rust_panic_with_hook::h310a921a123719cf
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/panicking.rs:799:13
  14:     0x75fca7c6c084 - std[8fa947ca3cb89c52]::panicking::begin_panic::<rustc_errors[dbe3df606ebeeaa3]::ExplicitBug>::{closure#0}
  15:     0x75fca7c68d06 - std[8fa947ca3cb89c52]::sys_common::backtrace::__rust_end_short_backtrace::<std[8fa947ca3cb89c52]::panicking::begin_panic<rustc_errors[dbe3df606ebeeaa3]::ExplicitBug>::{closure#0}, !>
  16:     0x75fca7c689e6 - std[8fa947ca3cb89c52]::panicking::begin_panic::<rustc_errors[dbe3df606ebeeaa3]::ExplicitBug>
  17:     0x75fca7c75261 - <rustc_errors[dbe3df606ebeeaa3]::diagnostic::BugAbort as rustc_errors[dbe3df606ebeeaa3]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  18:     0x75fca822e208 - <rustc_errors[dbe3df606ebeeaa3]::DiagCtxt>::span_bug::<rustc_span[28b37812f336c7b5]::span_encoding::Span, alloc[4c35cc17dd12f569]::string::String>
  19:     0x75fca824f6bd - rustc_middle[ec0047e274fe25be]::util::bug::opt_span_bug_fmt::<rustc_span[28b37812f336c7b5]::span_encoding::Span>::{closure#0}
  20:     0x75fca824f6ea - rustc_middle[ec0047e274fe25be]::ty::context::tls::with_opt::<rustc_middle[ec0047e274fe25be]::util::bug::opt_span_bug_fmt<rustc_span[28b37812f336c7b5]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x75fca824874b - rustc_middle[ec0047e274fe25be]::ty::context::tls::with_context_opt::<rustc_middle[ec0047e274fe25be]::ty::context::tls::with_opt<rustc_middle[ec0047e274fe25be]::util::bug::opt_span_bug_fmt<rustc_span[28b37812f336c7b5]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x75fca8247f67 - rustc_middle[ec0047e274fe25be]::util::bug::span_bug_fmt::<rustc_span[28b37812f336c7b5]::span_encoding::Span>
  23:     0x75fcaa92bdef - <rustc_const_eval[c0dfeea2f3bf4919]::interpret::eval_context::InterpCx<rustc_const_eval[c0dfeea2f3bf4919]::const_eval::dummy_machine::DummyMachine>>::overflowing_unary_op.cold
  24:     0x75fca67bda02 - <rustc_mir_transform[61a8aa7cac63b4d6]::known_panics_lint::ConstPropagator as rustc_middle[ec0047e274fe25be]::mir::visit::Visitor>::visit_assign
  25:     0x75fca993d83d - <rustc_mir_transform[61a8aa7cac63b4d6]::known_panics_lint::ConstPropagator as rustc_middle[ec0047e274fe25be]::mir::visit::Visitor>::visit_body
  26:     0x75fca6999566 - <rustc_mir_transform[61a8aa7cac63b4d6]::known_panics_lint::KnownPanicsLint as rustc_mir_transform[61a8aa7cac63b4d6]::pass_manager::MirLint>::run_lint
  27:     0x75fca90018cd - rustc_mir_transform[61a8aa7cac63b4d6]::pass_manager::run_passes_inner
  28:     0x75fca9395d42 - rustc_mir_transform[61a8aa7cac63b4d6]::mir_drops_elaborated_and_const_checked
  29:     0x75fca93955fd - rustc_query_impl[1888158ad4449b9b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[1888158ad4449b9b]::query_impl::mir_drops_elaborated_and_const_checked::dynamic_query::{closure#2}::{closure#0}, rustc_middle[ec0047e274fe25be]::query::erase::Erased<[u8; 8usize]>>
  30:     0x75fca938d171 - rustc_query_system[b5ee5054546adc93]::query::plumbing::try_execute_query::<rustc_query_impl[1888158ad4449b9b]::DynamicConfig<rustc_query_system[b5ee5054546adc93]::query::caches::VecCache<rustc_span[28b37812f336c7b5]::def_id::LocalDefId, rustc_middle[ec0047e274fe25be]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[1888158ad4449b9b]::plumbing::QueryCtxt, false>
  31:     0x75fca938cb2a - rustc_query_impl[1888158ad4449b9b]::query_impl::mir_drops_elaborated_and_const_checked::get_query_non_incr::__rust_end_short_backtrace
  32:     0x75fca936cff9 - rustc_interface[e7d426f81a3c67f]::passes::analysis
  33:     0x75fca936bddb - rustc_query_impl[1888158ad4449b9b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[1888158ad4449b9b]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[ec0047e274fe25be]::query::erase::Erased<[u8; 1usize]>>
  34:     0x75fca9dfb4e5 - rustc_query_system[b5ee5054546adc93]::query::plumbing::try_execute_query::<rustc_query_impl[1888158ad4449b9b]::DynamicConfig<rustc_query_system[b5ee5054546adc93]::query::caches::SingleCache<rustc_middle[ec0047e274fe25be]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[1888158ad4449b9b]::plumbing::QueryCtxt, false>
  35:     0x75fca9dfb249 - rustc_query_impl[1888158ad4449b9b]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  36:     0x75fca9c675ce - rustc_interface[e7d426f81a3c67f]::interface::run_compiler::<core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>, rustc_driver_impl[f52ddb602ce7d3c0]::run_compiler::{closure#0}>::{closure#1}
  37:     0x75fca9c54167 - std[8fa947ca3cb89c52]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[e7d426f81a3c67f]::util::run_in_thread_with_globals<rustc_interface[e7d426f81a3c67f]::util::run_in_thread_pool_with_globals<rustc_interface[e7d426f81a3c67f]::interface::run_compiler<core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>, rustc_driver_impl[f52ddb602ce7d3c0]::run_compiler::{closure#0}>::{closure#1}, core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>>::{closure#0}, core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>>
  38:     0x75fca9c53f2a - <<std[8fa947ca3cb89c52]::thread::Builder>::spawn_unchecked_<rustc_interface[e7d426f81a3c67f]::util::run_in_thread_with_globals<rustc_interface[e7d426f81a3c67f]::util::run_in_thread_pool_with_globals<rustc_interface[e7d426f81a3c67f]::interface::run_compiler<core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>, rustc_driver_impl[f52ddb602ce7d3c0]::run_compiler::{closure#0}>::{closure#1}, core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>>::{closure#0}, core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[19fe836d650da945]::result::Result<(), rustc_span[28b37812f336c7b5]::ErrorGuaranteed>>::{closure#2} as core[19fe836d650da945]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  39:     0x75fcab3c6cab - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h32bd23449113524a
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/alloc/src/boxed.rs:2022:9
  40:     0x75fcab3c6cab - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h04c486968e15e63c
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/alloc/src/boxed.rs:2022:9
  41:     0x75fcab3c6cab - std::sys::pal::unix::thread::Thread::new::thread_start::ha93fdc6af5a412fe
                               at /rustc/378a43a06510f3e3a49c69c8de71745e6a884048/library/std/src/sys/pal/unix/thread.rs:108:17
  42:     0x75fcab16555a - <unknown>
  43:     0x75fcab1e2a3c - <unknown>
  44:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.80.0-nightly (378a43a06 2024-05-01) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [mir_drops_elaborated_and_const_checked] elaborating drops for `main`
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.

```

</p>
</details>

<!--
query stack:
#0 [mir_drops_elaborated_and_const_checked] elaborating drops for `main`
#1 [analysis] running analysis passes on this crate

-->



<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"tgross35"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_const_eval/src/interpret/operator.rs
+++ b/compiler/rustc_const_eval/src/interpret/operator.rs
@@ -433,13 +433,16 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
             }
             ty::Float(fty) => {
                 let val = val.to_scalar();
+                if un_op != Neg {
+                    span_bug!(self.cur_span(), "Invalid float op {:?}", un_op);
+                }
+
                 // No NaN adjustment here, `-` is a bitwise operation!
-                let res = match (un_op, fty) {
-                    (Neg, FloatTy::F16) => Scalar::from_f16(-val.to_f16()?),
-                    (Neg, FloatTy::F32) => Scalar::from_f32(-val.to_f32()?),
-                    (Neg, FloatTy::F64) => Scalar::from_f64(-val.to_f64()?),
-                    (Neg, FloatTy::F128) => Scalar::from_f128(-val.to_f128()?),
-                    _ => span_bug!(self.cur_span(), "Invalid float op {:?}", un_op),
+                let res = match fty {
+                    FloatTy::F16 => Scalar::from_f16(-val.to_f16()?),
+                    FloatTy::F32 => Scalar::from_f32(-val.to_f32()?),
+                    FloatTy::F64 => Scalar::from_f64(-val.to_f64()?),
+                    FloatTy::F128 => Scalar::from_f128(-val.to_f128()?),
                 };
                 Ok(ImmTy::from_scalar(res, layout))
             }

--- a/compiler/rustc_const_eval/src/interpret/operator.rs
+++ b/compiler/rustc_const_eval/src/interpret/operator.rs
@@ -362,14 +362,18 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
                 let left = left.to_scalar();
                 let right = right.to_scalar();
                 Ok(match fty {
-                    FloatTy::F16 => unimplemented!("f16_f128"),
+                    FloatTy::F16 => {
+                        self.binary_float_op(bin_op, layout, left.to_f16()?, right.to_f16()?)
+                    }
                     FloatTy::F32 => {
                         self.binary_float_op(bin_op, layout, left.to_f32()?, right.to_f32()?)
                     }
                     FloatTy::F64 => {
                         self.binary_float_op(bin_op, layout, left.to_f64()?, right.to_f64()?)
                     }
-                    FloatTy::F128 => unimplemented!("f16_f128"),
+                    FloatTy::F128 => {
+                        self.binary_float_op(bin_op, layout, left.to_f128()?, right.to_f128()?)
+                    }
                 })
             }
             _ if left.layout.ty.is_integral() => {
@@ -431,8 +435,10 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
                 let val = val.to_scalar();
                 // No NaN adjustment here, `-` is a bitwise operation!
                 let res = match (un_op, fty) {
+                    (Neg, FloatTy::F16) => Scalar::from_f16(-val.to_f16()?),
                     (Neg, FloatTy::F32) => Scalar::from_f32(-val.to_f32()?),
                     (Neg, FloatTy::F64) => Scalar::from_f64(-val.to_f64()?),
+                    (Neg, FloatTy::F128) => Scalar::from_f128(-val.to_f128()?),
                     _ => span_bug!(self.cur_span(), "Invalid float op {:?}", un_op),
                 };
                 Ok(ImmTy::from_scalar(res, layout))

--- a/compiler/rustc_middle/src/mir/interpret/value.rs
+++ b/compiler/rustc_middle/src/mir/interpret/value.rs
@@ -69,6 +69,13 @@ impl<Prov: Provenance> fmt::LowerHex for Scalar<Prov> {
     }
 }
 
+impl<Prov> From<Half> for Scalar<Prov> {
+    #[inline(always)]
+    fn from(f: Half) -> Self {
+        Scalar::from_f16(f)
+    }
+}
+
 impl<Prov> From<Single> for Scalar<Prov> {
     #[inline(always)]
     fn from(f: Single) -> Self {
@@ -80,6 +87,13 @@ impl<Prov> From<Double> for Scalar<Prov> {
     #[inline(always)]
     fn from(f: Double) -> Self {
         Scalar::from_f64(f)
+    }
+}
+
+impl<Prov> From<Quad> for Scalar<Prov> {
+    #[inline(always)]
+    fn from(f: Quad) -> Self {
+        Scalar::from_f128(f)
     }
 }
 

--- a/src/tools/miri/src/helpers.rs
+++ b/src/tools/miri/src/helpers.rs
@@ -7,7 +7,7 @@ use std::time::Duration;
 
 use rand::RngCore;
 
-use rustc_apfloat::ieee::{Double, Single};
+use rustc_apfloat::ieee::{Double, Half, Quad, Single};
 use rustc_apfloat::Float;
 use rustc_hir::{
     def::{DefKind, Namespace},
@@ -1201,12 +1201,14 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         };
 
         let (val, status) = match fty {
-            FloatTy::F16 => unimplemented!("f16_f128"),
+            FloatTy::F16 =>
+                float_to_int_inner::<Half>(this, src.to_scalar().to_f16()?, cast_to, round),
             FloatTy::F32 =>
                 float_to_int_inner::<Single>(this, src.to_scalar().to_f32()?, cast_to, round),
             FloatTy::F64 =>
                 float_to_int_inner::<Double>(this, src.to_scalar().to_f64()?, cast_to, round),
-            FloatTy::F128 => unimplemented!("f16_f128"),
+            FloatTy::F128 =>
+                float_to_int_inner::<Quad>(this, src.to_scalar().to_f128()?, cast_to, round),
         };
 
         if status.intersects(

--- a/src/tools/miri/tests/pass/float.rs
+++ b/src/tools/miri/tests/pass/float.rs
@@ -1,6 +1,8 @@
 #![feature(stmt_expr_attributes)]
 #![feature(float_gamma)]
 #![feature(core_intrinsics)]
+#![feature(f128)]
+#![feature(f16)]
 #![allow(arithmetic_overflow)]
 
 use std::fmt::Debug;
@@ -41,103 +43,23 @@ trait FloatToInt<Int>: Copy {
     unsafe fn cast_unchecked(self) -> Int;
 }
 
-impl FloatToInt<i8> for f32 {
-    fn cast(self) -> i8 {
-        self as _
-    }
-    unsafe fn cast_unchecked(self) -> i8 {
-        self.to_int_unchecked()
-    }
-}
-impl FloatToInt<i32> for f32 {
-    fn cast(self) -> i32 {
-        self as _
-    }
-    unsafe fn cast_unchecked(self) -> i32 {
-        self.to_int_unchecked()
-    }
-}
-impl FloatToInt<u32> for f32 {
-    fn cast(self) -> u32 {
-        self as _
-    }
-    unsafe fn cast_unchecked(self) -> u32 {
-        self.to_int_unchecked()
-    }
-}
-impl FloatToInt<i64> for f32 {
-    fn cast(self) -> i64 {
-        self as _
-    }
-    unsafe fn cast_unchecked(self) -> i64 {
-        self.to_int_unchecked()
-    }
-}
-impl FloatToInt<u64> for f32 {
-    fn cast(self) -> u64 {
-        self as _
-    }
-    unsafe fn cast_unchecked(self) -> u64 {
-        self.to_int_unchecked()
-    }
+macro_rules! float_to_int {
+    ($fty:ty => $($ity:ty),+ $(,)?) => {
+        $(
+            impl FloatToInt<$ity> for $fty {
+                fn cast(self) -> $ity {
+                    self as _
+                }
+                unsafe fn cast_unchecked(self) -> $ity {
+                    self.to_int_unchecked()
+                }
+            }
+        )*
+    };
 }
 
-impl FloatToInt<i8> for f64 {
-    fn cast(self) -> i8 {
-        self as _
-    }
-    unsafe fn cast_unchecked(self) -> i8 {
-        self.to_int_unchecked()
-    }
-}
-impl FloatToInt<i32> for f64 {
-    fn cast(self) -> i32 {
-        self as _
-    }
-    unsafe fn cast_unchecked(self) -> i32 {
-        self.to_int_unchecked()
-    }
-}
-impl FloatToInt<u32> for f64 {
-    fn cast(self) -> u32 {
-        self as _
-    }
-    unsafe fn cast_unchecked(self) -> u32 {
-        self.to_int_unchecked()
-    }
-}
-impl FloatToInt<i64> for f64 {
-    fn cast(self) -> i64 {
-        self as _
-    }
-    unsafe fn cast_unchecked(self) -> i64 {
-        self.to_int_unchecked()
-    }
-}
-impl FloatToInt<u64> for f64 {
-    fn cast(self) -> u64 {
-        self as _
-    }
-    unsafe fn cast_unchecked(self) -> u64 {
-        self.to_int_unchecked()
-    }
-}
-impl FloatToInt<i128> for f64 {
-    fn cast(self) -> i128 {
-        self as _
-    }
-    unsafe fn cast_unchecked(self) -> i128 {
-        self.to_int_unchecked()
-    }
-}
-impl FloatToInt<u128> for f64 {
-    fn cast(self) -> u128 {
-        self as _
-    }
-    unsafe fn cast_unchecked(self) -> u128 {
-        self.to_int_unchecked()
-    }
-}
+float_to_int!(f32 => i8, u8, i16, u16, i32, u32, i64, u64, i128, u128);
+float_to_int!(f64 => i8, u8, i16, u16, i32, u32, i64, u64, i128, u128);
 
 /// Test this cast both via `as` and via `approx_unchecked` (i.e., it must not saturate).
 #[track_caller]
@@ -153,18 +75,29 @@ where
 
 fn basic() {
     // basic arithmetic
+    assert_eq(6.0_f16 * 6.0_f16, 36.0_f16);
     assert_eq(6.0_f32 * 6.0_f32, 36.0_f32);
     assert_eq(6.0_f64 * 6.0_f64, 36.0_f64);
+    assert_eq(6.0_f128 * 6.0_f128, 36.0_f128);
+    assert_eq(-{ 5.0_f16 }, -5.0_f16);
     assert_eq(-{ 5.0_f32 }, -5.0_f32);
     assert_eq(-{ 5.0_f64 }, -5.0_f64);
+    assert_eq(-{ 5.0_f128 }, -5.0_f128);
+
     // infinities, NaN
+    // FIXME(f16_f128): add when constants and `is_infinite` are available
     assert!((5.0_f32 / 0.0).is_infinite());
     assert_ne!({ 5.0_f32 / 0.0 }, { -5.0_f32 / 0.0 });
     assert!((5.0_f64 / 0.0).is_infinite());
     assert_ne!({ 5.0_f64 / 0.0 }, { 5.0_f64 / -0.0 });
     assert_ne!(f32::NAN, f32::NAN);
     assert_ne!(f64::NAN, f64::NAN);
+
     // negative zero
+    let posz = 0.0f16;
+    let negz = -0.0f16;
+    assert_eq(posz, negz);
+    assert_ne!(posz.to_bits(), negz.to_bits());
     let posz = 0.0f32;
     let negz = -0.0f32;
     assert_eq(posz, negz);
@@ -173,15 +106,30 @@ fn basic() {
     let negz = -0.0f64;
     assert_eq(posz, negz);
     assert_ne!(posz.to_bits(), negz.to_bits());
+    let posz = 0.0f128;
+    let negz = -0.0f128;
+    assert_eq(posz, negz);
+    assert_ne!(posz.to_bits(), negz.to_bits());
+
     // byte-level transmute
-    let x: u64 = unsafe { std::mem::transmute(42.0_f64) };
-    let y: f64 = unsafe { std::mem::transmute(x) };
-    assert_eq(y, 42.0_f64);
+    let x: u16 = unsafe { std::mem::transmute(42.0_f16) };
+    let y: f16 = unsafe { std::mem::transmute(x) };
+    assert_eq(y, 42.0_f16);
     let x: u32 = unsafe { std::mem::transmute(42.0_f32) };
     let y: f32 = unsafe { std::mem::transmute(x) };
     assert_eq(y, 42.0_f32);
+    let x: u64 = unsafe { std::mem::transmute(42.0_f64) };
+    let y: f64 = unsafe { std::mem::transmute(x) };
+    assert_eq(y, 42.0_f64);
+    let x: u128 = unsafe { std::mem::transmute(42.0_f128) };
+    let y: f128 = unsafe { std::mem::transmute(x) };
+    assert_eq(y, 42.0_f128);
 
     // `%` sign behavior, some of this used to be buggy
+    assert!((black_box(1.0f16) % 1.0).is_sign_positive());
+    assert!((black_box(1.0f16) % -1.0).is_sign_positive());
+    assert!((black_box(-1.0f16) % 1.0).is_sign_negative());
+    assert!((black_box(-1.0f16) % -1.0).is_sign_negative());
     assert!((black_box(1.0f32) % 1.0).is_sign_positive());
     assert!((black_box(1.0f32) % -1.0).is_sign_positive());
     assert!((black_box(-1.0f32) % 1.0).is_sign_negative());
@@ -190,7 +138,12 @@ fn basic() {
     assert!((black_box(1.0f64) % -1.0).is_sign_positive());
     assert!((black_box(-1.0f64) % 1.0).is_sign_negative());
     assert!((black_box(-1.0f64) % -1.0).is_sign_negative());
+    assert!((black_box(1.0f128) % 1.0).is_sign_positive());
+    assert!((black_box(1.0f128) % -1.0).is_sign_positive());
+    assert!((black_box(-1.0f128) % 1.0).is_sign_negative());
+    assert!((black_box(-1.0f128) % -1.0).is_sign_negative());
 
+    // FIXME(f16_f128): add when `abs` is available
     assert_eq!((-1.0f32).abs(), 1.0f32);
     assert_eq!(34.2f64.abs(), 34.2f64);
 }

--- a/tests/crashes/124583.rs
+++ b/tests/crashes/124583.rs
@@ -1,5 +1,0 @@
-//@ known-bug: rust-lang/rust#124583
-
-fn main() {
-    let _ = -(-0.0f16);
-}

--- a/tests/ui/numbers-arithmetic/f16-f128-lit.rs
+++ b/tests/ui/numbers-arithmetic/f16-f128-lit.rs
@@ -1,3 +1,5 @@
+// Make sure negation happens correctly. Also included:
+// issue: rust-lang/rust#124583
 //@ run-pass
 
 #![feature(f16)]
@@ -8,9 +10,11 @@ fn main() {
     assert_eq!((-0.0_f16).to_bits(), 0x8000);
     assert_eq!(10.0_f16.to_bits(), 0x4900);
     assert_eq!((-10.0_f16).to_bits(), 0xC900);
+    assert_eq!((-(-0.0f16)).to_bits(), 0x0000);
 
     assert_eq!(0.0_f128.to_bits(), 0x0000_0000_0000_0000_0000_0000_0000_0000);
     assert_eq!((-0.0_f128).to_bits(), 0x8000_0000_0000_0000_0000_0000_0000_0000);
     assert_eq!(10.0_f128.to_bits(), 0x4002_4000_0000_0000_0000_0000_0000_0000);
     assert_eq!((-10.0_f128).to_bits(), 0xC002_4000_0000_0000_0000_0000_0000_0000);
+    assert_eq!((-(-0.0f128)).to_bits(), 0x0000_0000_0000_0000_0000_0000_0000_0000);
 }

